File: merge_restaurants.js
/*
  Requirements:
    npm install xlsx

  Usage:
    node "merge_restaurants.js"
*/

const fs = require('fs');
const path = require('path');
const XLSX = require('xlsx');

const BASE_DIR = __dirname;

const OUT_COLUMNS = [
  'RestaurantName',
  'Category',
  'ItemName',
  'Price',
  'Size / Variant',
  'AdditionalInfo',
];

const SYNONYMS = {
  Category: [
    'category',
    'type',
    'segment',
    'menu category',
    'group',
    'section',
  ],
  ItemName: [
    'item',
    'item name',
    'product',
    'product name',
    'menu item',
    'dish',
    'name',
    'pizza',
    'burger',
  ],
  Price: [
    'price',
    'rate',
    'amount',
    'cost',
    'mrp',
    'rs',
    'rupees',
  ],
  'Size / Variant': [
    'size',
    'variant',
    'portion',
    'crust',
    'topping',
    'weight',
    'ml',
    'pack',
    'medium/large',
  ],
  AdditionalInfo: [
    'description',
    'notes',
    'detail',
    'details',
    'info',
    'remarks',
  ],
};

const RESTAURANT_NAME_MAP = new Map([
  ['mcdonalds', "McDonald's"],
  ['pizza_hut', 'Pizza Hut'],
  ['pizzahut', 'Pizza Hut'],
  ['dominos', "Domino's"],
  ['burger_king', 'Burger King'],
  ['burgerking', 'Burger King'],
]);

const INPUT_FILES = [
  path.join(BASE_DIR, 'McDonalds_Price_Comparison_2025.xlsx'),
  path.join(BASE_DIR, 'pizza_hut_price_comparison.xlsx'),
  path.join(BASE_DIR, 'dominos_price_comparison.xlsx'),
];
// Detect Burger King file dynamically (.xlsx or .csv)
(() => {
  for (const ext of ['.xlsx', '.csv']) {
    const candidate = path.join(BASE_DIR, `burger_king_price_comparison${ext}`);
    if (fs.existsSync(candidate)) {
      INPUT_FILES.push(candidate);
      break;
    }
  }
})();

const OUTPUT_XLSX = path.join(BASE_DIR, 'restaurants.xlsx');
const OUTPUT_CSV = path.join(BASE_DIR, 'restaurants_merged.csv');

function _norm(s) {
  return String(s).trim().toLowerCase().replace(/\s+/g, ' ');
}

function _findBestMatch(columns, targets) {
  const normCols = new Map(columns.map((c) => [c, _norm(c)]));
  for (const t of targets) {
    const tNorm = _norm(t);
    // exact
    for (const [c, n] of normCols.entries()) {
      if (n === tNorm) return c;
    }
    // startswith / contains
    for (const [c, n] of normCols.entries()) {
      if (n.startsWith(tNorm) || n.includes(tNorm)) return c;
    }
  }
  return null;
}

function _mapColumns(rows) {
  if (!rows || rows.length === 0) return [];
  const columns = Object.keys(rows[0] || {});
  const mapping = {};
  for (const [target, variants] of Object.entries(SYNONYMS)) {
    mapping[target] = _findBestMatch(columns, variants);
  }
  return rows.map((r) => ({
    Category: mapping['Category'] ? r[mapping['Category']] : null,
    ItemName: mapping['ItemName'] ? r[mapping['ItemName']] : null,
    Price: mapping['Price'] ? r[mapping['Price']] : null,
    'Size / Variant': mapping['Size / Variant'] ? r[mapping['Size / Variant']] : null,
    AdditionalInfo: mapping['AdditionalInfo'] ? r[mapping['AdditionalInfo']] : null,
  }));
}

function _toPrice(val) {
  if (val === undefined || val === null || val === '') return null;
  let s = String(val);
  s = s.replace(/,/g, '');
  s = s.replace(/[^0-9.]+/g, '');
  if (!s) return null;
  const num = Number(s);
  return Number.isFinite(num) ? num : null;
}

function _cleanRows(rows, restaurantName) {
  const cleaned = [];
  for (const row of rows) {
    const obj = { ...row };
    for (const key of ['Category', 'ItemName', 'Size / Variant', 'AdditionalInfo']) {
      const v = obj[key];
      if (v === undefined || v === null) {
        obj[key] = null;
      } else if (typeof v === 'string') {
        const trimmed = v.trim();
        obj[key] = trimmed.length ? trimmed : null;
      } else {
        obj[key] = v;
      }
    }
    obj.Price = _toPrice(obj.Price);
    obj.RestaurantName = restaurantName;
    cleaned.push({
      RestaurantName: obj.RestaurantName,
      Category: obj.Category ?? null,
      ItemName: obj.ItemName ?? null,
      Price: obj.Price ?? null,
      'Size / Variant': obj['Size / Variant'] ?? null,
      AdditionalInfo: obj.AdditionalInfo ?? null,
    });
  }
  // Drop fully empty rows for key columns
  const filtered = cleaned.filter((r) => {
    return (
      (r.ItemName && r.ItemName !== '') ||
      (r.Price !== null && r.Price !== undefined) ||
      (r.Category && r.Category !== '') ||
      (r['Size / Variant'] && r['Size / Variant'] !== '')
    );
  });

  // Remove duplicates (composite key)
  const seen = new Set();
  const unique = [];
  for (const r of filtered) {
    const key = [
      r.RestaurantName ?? '',
      r.Category ?? '',
      r.ItemName ?? '',
      r['Size / Variant'] ?? '',
      r.Price ?? ''
    ].join('|\u241F|');
    if (!seen.has(key)) {
      seen.add(key);
      unique.push(r);
    }
  }
  return unique;
}

function _restaurantFromPath(p) {
  const stem = path.basename(p, path.extname(p)).toLowerCase();
  for (const [key, val] of RESTAURANT_NAME_MAP.entries()) {
    if (stem.includes(key)) return val;
  }
  return stem.replace(/_/g, ' ').replace(/\b\w/g, (m) => m.toUpperCase());
}

function _readAll() {
  const perRestaurant = new Map();
  for (const file of INPUT_FILES) {
    if (!fs.existsSync(file)) {
      console.warn(`Warning: file not found: ${file}`);
      continue;
    }
    const rname = _restaurantFromPath(file);
    const ext = path.extname(file).toLowerCase();
    const frames = [];

    try {
      const wb = XLSX.readFile(file, { cellDates: false, raw: false });
      const sheetNames = wb.SheetNames;
      for (const s of sheetNames) {
        const ws = wb.Sheets[s];
        const json = XLSX.utils.sheet_to_json(ws, { defval: null });
        frames.push(json);
      }
    } catch (err) {
      console.error(`Failed reading ${file}:`, err.message);
    }

    if (frames.length) {
      if (!perRestaurant.has(rname)) perRestaurant.set(rname, []);
      perRestaurant.get(rname).push(...frames);
    }
  }
  return perRestaurant;
}

function writeExcel(perRestaurantClean) {
  const wb = XLSX.utils.book_new();
  for (const [rname, rows] of perRestaurantClean.entries()) {
    const aoa = [OUT_COLUMNS];
    for (const r of rows) {
      aoa.push([
        r.RestaurantName,
        r.Category,
        r.ItemName,
        r.Price,
        r['Size / Variant'],
        r.AdditionalInfo,
      ]);
    }
    const ws = XLSX.utils.aoa_to_sheet(aoa);
    const sheetName = rname.replace(/[\\/*?:\[\]]/g, ' ').slice(0, 31) || 'Sheet';
    XLSX.utils.book_append_sheet(wb, ws, sheetName);
  }
  XLSX.writeFile(wb, OUTPUT_XLSX);
  console.log(`Wrote Excel: ${OUTPUT_XLSX}`);
}

function writeCsv(mergedAll) {
  const wb = XLSX.utils.book_new();
  const aoa = [OUT_COLUMNS];
  for (const r of mergedAll) {
    aoa.push([
      r.RestaurantName,
      r.Category,
      r.ItemName,
      r.Price,
      r['Size / Variant'],
      r.AdditionalInfo,
    ]);
  }
  const ws = XLSX.utils.aoa_to_sheet(aoa);
  XLSX.utils.book_append_sheet(wb, ws, 'All');
  const csv = XLSX.utils.sheet_to_csv(ws);
  fs.writeFileSync(OUTPUT_CSV, csv, 'utf8');
  console.log(`Wrote CSV: ${OUTPUT_CSV}`);
}

function main() {
  const perRestaurantRaw = _readAll();
  const perRestaurantClean = new Map();
  const mergedAll = [];

  for (const [rname, frames] of perRestaurantRaw.entries()) {
    const mapped = frames.map((rows) => _mapColumns(rows));
    const flat = mapped.flat();
    if (!flat.length) continue;
    const cleaned = _cleanRows(flat, rname);
    if (!cleaned.length) continue;
    perRestaurantClean.set(rname, cleaned);
    mergedAll.push(...cleaned);
  }

  if (!mergedAll.length) {
    console.log('No data loaded. Please verify input files.');
    return;
  }

  writeCsv(mergedAll);
  writeExcel(perRestaurantClean);
}

if (require.main === module) {
  main();
}

File: merge_restaurants.py
from __future__ import annotations
import re
from pathlib import Path
from typing import Dict, List, Optional

import pandas as pd

# Requirements:
#   pip install pandas openpyxl

# Normalized output columns
OUT_COLUMNS = [
    "RestaurantName",
    "Category",
    "ItemName",
    "Price",
    "Size / Variant",
    "AdditionalInfo",
]

# Column synonym map: normalized target -> list of possible column name variants
SYNONYMS: Dict[str, List[str]] = {
    "Category": [
        "category",
        "type",
        "segment",
        "menu category",
        "group",
        "section",
    ],
    "ItemName": [
        "item",
        "item name",
        "product",
        "product name",
        "menu item",
        "dish",
        "name",
        "pizza",
        "burger",
    ],
    "Price": [
        "price",
        "rate",
        "amount",
        "cost",
        "mrp",
        "rs",
        "rupees",
    ],
    "Size / Variant": [
        "size",
        "variant",
        "portion",
        "crust",
        "topping",
        "weight",
        "ml",
        "pack",
        "medium/large",
    ],
    "AdditionalInfo": [
        "description",
        "notes",
        "detail",
        "details",
        "info",
        "remarks",
    ],
}

RESTAURANT_NAME_MAP = {
    "mcdonalds": "McDonald's",
    "pizza_hut": "Pizza Hut",
    "pizzahut": "Pizza Hut",
    "dominos": "Domino's",
    "burger_king": "Burger King",
    "burgerking": "Burger King",
}

BASE_DIR = Path(__file__).parent

INPUT_FILES = [
    BASE_DIR / "McDonalds_Price_Comparison_2025.xlsx",
    BASE_DIR / "pizza_hut_price_comparison.xlsx",
    BASE_DIR / "dominos_price_comparison.xlsx",
    # Burger King may be xlsx or csv; we will detect below
]

# Detect Burger King file dynamically
for ext in (".xlsx", ".csv"):
    candidate = BASE_DIR / f"burger_king_price_comparison{ext}"
    if candidate.exists():
        INPUT_FILES.append(candidate)
        break

OUTPUT_XLSX = BASE_DIR / "restaurants.xlsx"
OUTPUT_CSV = BASE_DIR / "restaurants_merged.csv"


def _normalize_col(s: str) -> str:
    return re.sub(r"\s+", " ", s.strip().lower())


def _find_best_match(columns: List[str], targets: List[str]) -> Optional[str]:
    norm_cols = {c: _normalize_col(c) for c in columns}
    for t in targets:
        t_norm = _normalize_col(t)
        # Exact match first
        for c, n in norm_cols.items():
            if n == t_norm:
                return c
        # Startswith/contains fallback
        for c, n in norm_cols.items():
            if n.startswith(t_norm) or t_norm in n:
                return c
    return None


def _map_columns(df: pd.DataFrame) -> pd.DataFrame:
    if df is None or df.empty:
        return pd.DataFrame(columns=OUT_COLUMNS)

    df = df.copy()
    df.columns = [str(c) for c in df.columns]

    # Build detected mapping
    mapping: Dict[str, Optional[str]] = {}
    for target, variants in SYNONYMS.items():
        src = _find_best_match(list(df.columns), variants)
        mapping[target] = src

    # Construct normalized frame
    out = pd.DataFrame()
    out["Category"] = df[mapping["Category"]] if mapping["Category"] in df else None
    out["ItemName"] = df[mapping["ItemName"]] if mapping["ItemName"] in df else None
    out["Price"] = df[mapping["Price"]] if mapping["Price"] in df else None
    out["Size / Variant"] = (
        df[mapping["Size / Variant"]] if mapping["Size / Variant"] in df else None
    )
    out["AdditionalInfo"] = (
        df[mapping["AdditionalInfo"]] if mapping["AdditionalInfo"] in df else None
    )

    return out


def _clean_frame(df: pd.DataFrame, restaurant_name: str) -> pd.DataFrame:
    if df is None or df.empty:
        return pd.DataFrame(columns=OUT_COLUMNS)

    df = df.copy()

    # Trim whitespace in string-like columns
    for col in ["Category", "ItemName", "Size / Variant", "AdditionalInfo"]:
        if col in df:
            df[col] = df[col].astype(str).where(df[col].notna(), None)
            df[col] = df[col].apply(lambda x: x.strip() if isinstance(x, str) else x)
            # Replace empty strings with None
            df[col] = df[col].apply(lambda x: None if isinstance(x, str) and x == "" else x)

    # Normalize price: strip currency symbols and commas; coerce to float
    if "Price" in df:
        def to_price(val):
            if pd.isna(val):
                return pd.NA
            s = str(val)
            s = s.replace(",", "")
            s = re.sub(r"[^0-9.]+", "", s)
            try:
                return float(s) if s else pd.NA
            except Exception:
                return pd.NA
        df["Price"] = df["Price"].apply(to_price)

    # Add RestaurantName and re-order
    df["RestaurantName"] = restaurant_name
    df = df[[
        "RestaurantName",
        "Category",
        "ItemName",
        "Price",
        "Size / Variant",
        "AdditionalInfo",
    ]]

    # Drop fully empty rows for key columns
    df = df.dropna(how="all", subset=["ItemName", "Price", "Category", "Size / Variant"])

    # Remove duplicates
    df = df.drop_duplicates(subset=[
        "RestaurantName",
        "Category",
        "ItemName",
        "Size / Variant",
        "Price",
    ])

    return df


def _restaurant_from_path(path: Path) -> str:
    stem = path.stem.lower()
    for key, name in RESTAURANT_NAME_MAP.items():
        if key in stem:
            return name
    # Fallback to title-case of filename
    return stem.replace("_", " ").title()


def _load_all() -> Dict[str, List[pd.DataFrame]]:
    per_restaurant: Dict[str, List[pd.DataFrame]] = {}
    for path in INPUT_FILES:
        if not path.exists():
            print(f"Warning: file not found: {path}")
            continue
        rname = _restaurant_from_path(path)
        frames: List[pd.DataFrame] = []
        if path.suffix.lower() == ".csv":
            try:
                df = pd.read_csv(path)
                frames.append(df)
            except Exception as e:
                print(f"Failed reading CSV {path}: {e}")
        else:
            try:
                sheets = pd.read_excel(path, sheet_name=None)
                for _, df in sheets.items():
                    frames.append(df)
            except Exception as e:
                print(f"Failed reading Excel {path}: {e}")
        if frames:
            per_restaurant.setdefault(rname, []).extend(frames)
    return per_restaurant


def main() -> None:
    per_restaurant_raw = _load_all()

    per_restaurant_clean: Dict[str, pd.DataFrame] = {}
    merged_frames: List[pd.DataFrame] = []

    for rname, frames in per_restaurant_raw.items():
        normalized_dfs = [_map_columns(df) for df in frames]
        if not normalized_dfs:
            continue
        combined = pd.concat(normalized_dfs, ignore_index=True)
        cleaned = _clean_frame(combined, rname)
        if cleaned is None or cleaned.empty:
            continue
        per_restaurant_clean[rname] = cleaned
        merged_frames.append(cleaned)

    if not merged_frames:
        print("No data loaded. Please verify input files.")
        return

    merged_all = pd.concat(merged_frames, ignore_index=True)

    # Export CSV (database-ready)
    merged_all.to_csv(OUTPUT_CSV, index=False)
    print(f"Wrote CSV: {OUTPUT_CSV}")

    # Export Excel with one sheet per restaurant
    with pd.ExcelWriter(OUTPUT_XLSX, engine="openpyxl") as writer:
        for rname, df in per_restaurant_clean.items():
            # Excel sheet names max length 31, remove invalid chars
            sheet = re.sub(r"[\\/*?:\[\]]", " ", rname)[:31]
            df.to_excel(writer, sheet_name=sheet, index=False)
    print(f"Wrote Excel: {OUTPUT_XLSX}")


if __name__ == "__main__":
    main()
